### ISSUE TITLE
pyright 1.1.405 regression - `reportUnnecessaryComparison` false positive under extremely specific circumstances when re-assigning a variable in a loop using its previous value

### ISSUE THREAD

**Issue by DetachHead** [DetachHead/basedpyright#1457]
### Description

Code sample in [basedpyright playground](https://basedpyright.com/?pyrightVersion=1.31.4&typeCheckingMode=all&reportUnusedParameter=false&code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwFgAoUSKJAT1y1IHNseAkQAyWJGhDEANiznMAxtOL16UAIIBtACpQAvFDESp0zQEYANFABMAXVsAuFlBdQAJmmCwYvcXAAU9GjSwFYAbjIIaA5Q2gCUTsyuyVBBIQB0EdJRMboGWVHyLEoqagBCxCA6js6uHl7ApGaBwcAxFVXatglQ6X3yyfVQjdb%2BtSmprRbjKZV8ZjHSWPRImrxIttNJEyQgfNYx6lvJCTPJ6PQI0kR6ZmeupGgA7jAF0Rr6u-v3LsAEsNhSFBzDVtjtXBcrjcoHcweDsF45gsfvC%2BukUcksF5Hi83plImh9AYDhidmjSS4ca8CZ91P4YQioFS8W8iQzgkEbHEoABiKCScAgGLofCEACq5DQCjQqkqnAAwpBcJVlnQUegUCAgZDrkA)

```python
from __future__ import annotations
from typing import Literal


class A[T = Literal[1, 2]]:
    def __init__(self, value: T):
        self.value: T = value


class Bar[T]:
    def fn1(self: Bar[T]): ...

    def fn2(
        self,
        arg1: list[int],
        arg2: A,
    ):
        result =1
        new_value: A = arg2
        for _ in [1]:
            result = 1
            if arg1:
                ...
            if new_value.value == 2:
                ...
            new_value = A( 1 if new_value.value == 1 else 2) # error: reportUnnecessaryComparison
        return result
```

this is as minimal as i can get it, removing any seeminly unrelated part of this repro causes the error to go away. 

related: #1456
